### PR TITLE
[Trading 3/5] Batched bar protocol + symbol-indexed order book (#377)

### PR DESCRIPTION
**Pull request by brandonkindred**
## Summary

Implements [#377](https://github.com/brandonkindred/Khala-Agentic-AI-Teams/issues/377) — two coordinated changes that ship together:

1. **Batched subprocess bar protocol** — a one-year 15-minute × 10-symbol backtest no longer pays ~30 ms × 250k bars in subprocess round-trip overhead.
2. **Symbol-indexed order book with O(1) cancel** — `cancel`/`remove` no longer pay an O(n) `list.remove()` on the per-symbol bucket, so per-bar order routing stays O(active-for-symbol) under the new throughput.

### OrderBook
- `_by_symbol: Dict[str, deque[str]]` with **lazy tombstoning**: cancel only pops from `_pending` (O(1)); the deque entry is filtered out lazily by `pending_for_symbol`, which triggers an amortized rebuild once tombstones reach parity with live entries.
- New `_by_parent: Dict[str, deque[str]]` index gives `children_of` and `_cascade_cancel_children` the same O(1) treatment so the cancel cascade isn't silently O(N) via a `_pending.values()` scan.
- 6 new tests in `test_order_book.py` covering tombstoning, FIFO order preservation, no-eager-compaction, and a quadratic-regression guard.

### Streaming harness
- Capability handshake: child advertises `{"chunked_bars": true}` in every `ready`. Parent exposes `StreamingHarness.supports_chunked_bars` and falls back to per-bar mode silently (with a logged warning) when the child predates capability negotiation.
- New `send_bars(bars=[...])` with a `{"kind": "bars", ...}` message.
- Each emitted `order` / `cancel` carries a `bar_index` field tying it to the source bar within the chunk, so per-order `submitted_at` preserves `BarSafetyAssertion` semantics under chunked mode.
- `HarnessResponse` gains parallel `order_bar_indices` / `cancel_bar_indices` lists + a `capabilities` dict.

### Strategy contract
- New `Strategy.on_bars(ctx, bars)` default that delegates to `on_bar` per bar. The child harness detects the override (via `type(instance).on_bars is not contract.Strategy.on_bars`) and dispatches to the vectorised path when present, else loops `on_bar` with `bar_index` tagging.
- `StrategyContext` gains private `_current_bar_index` consumed by `submit_order` / `cancel` to attach `bar_index` when set.

### Trading service
- New `BAR_CHUNK_SIZE` env (default `1`, paper-trade pinned to `1` via `bar_chunk_size=1` in `__init__`) plus `_resolve_bar_chunk_size()`.
- New `TradingService._run_chunked` path: buffers up to `chunk_size` bars, sends them via `send_bars`, then replays each bar's pre-/post-bar logic in order using the `bar_index`-tagged response. Runs only when `BAR_CHUNK_SIZE > 1` **AND** the child advertises `chunked_bars`; default behaviour is byte-identical to the existing per-bar path.
- Tradeoff: every bar in a chunk sees the chunk-start state snapshot — state-dependent strategies should keep `BAR_CHUNK_SIZE=1`. Paper trading is pinned for that reason.

### Tests
- `test_streaming_harness_chunked.py`: capability negotiation, `bar_index` round-trip, `on_bars` override dispatch, empty-chunk edge case, per-bar-path-still-works, and a service-level chunked run with the round-trip strategy that asserts no `lookahead_violation` under `BAR_CHUNK_SIZE=4`.
- `bench/bench_intraday_15m.py` + `bench` marker registration in `conftest.py`: 24/7 synthetic 15-min crypto fixture; asserts ≥2× speedup in CI on a ~10k-event fixture (production target 10× lives behind `BENCH_INTRADAY_FULL=1` and assumes the sandboxed-subprocess per-round-trip cost the issue's analysis was based on; on local subprocess we measure ~2.7×).

All 239 tests in the touched suites pass. Golden parity at `BAR_CHUNK_SIZE=1` holds trivially since the chunked path only fires when the env requests it.

## Test plan

- [x] `pytest backend/agents/investment_team/tests/test_order_book.py` — 117 pass (111 existing + 6 new)
- [x] `pytest backend/agents/investment_team/tests/test_streaming_harness_chunked.py` — 6 pass
- [x] `pytest backend/agents/investment_team/tests/test_bar_safety.py` — 12 pass
- [x] `pytest backend/agents/investment_team/tests/golden/test_simulator_invariants.py` — 4 pass (golden parity at default `BAR_CHUNK_SIZE=1`)
- [x] `pytest backend/agents/investment_team/tests/test_paper_trade.py` — 12 pass (paper-trade still uses per-bar path via `bar_chunk_size=1`)
- [x] `pytest backend/agents/investment_team/tests/test_trading_service.py` `test_partial_fills.py` `test_realistic_fill_model.py` `test_contract_gates.py` — 88 pass
- [x] `pytest -m bench backend/agents/investment_team/tests/bench/bench_intraday_15m.py` — passes 2× threshold (measured 2.7× on local subprocess)
- [x] `python -m ruff check` and `python -m ruff format --check` — clean
- [ ] Smoke test under sandboxed-subprocess deployment to verify the production 10× target

https://claude.ai/code/session_013LaFiHiTVGzWTarDmhgkTP

---
_Generated by [Claude Code](https://claude.ai/code/session_013LaFiHiTVGzWTarDmhgkTP)_

### CODE REVIEW

File: backend/agents/investment_team/trading_service/strategy/streaming_harness.py [hunk 1]
```diff
@@ -342,6 +402,45 @@ def main():
...
+                        bars = []
+                        for item in chunk:
+                            bar = contract.Bar(**item["bar"])
+                            state = item.get("state") or {}
+                            _apply_state(
+                                ctx, state, is_warmup=bool(item.get("is_warmup", False))
+                            )
+                            ctx._ingest_bar(bar)
+                            bars.append(bar)
+                        instance.on_bars(ctx, bars)
```
[the remark this answers is not in the document]
Comment (brandonkindred) [reply backend/agents/investment_team/trading_service/strategy/streaming_harness.py]: Addressed in a756d8e by **rejecting** the `on_bars` override outright under the chunked protocol rather than trying to constrain it. The child harness now detects `type(instance).on_bars is not contract.Strategy.on_bars` and emits a structured `contract_error` pointing authors at `on_bar` or `BAR_CHUNK_SIZE=1`. Constraining the override (e.g. forcing `submitted_at = chunk_end`, or requiring monotonically-non-decreasing `bar_index`) doesn't actually close the hole — the strategy still has access to all bars in the chunk and can use later-bar data to inform an earlier-bar emission. Outright rejection is the only fix that preserves the look-ahead safety model the rest of the engine relies on.

New regression test: `test_on_bars_override_is_rejected_under_chunked_protocol` asserts the `contract_error` etype and the `BAR_CHUNK_SIZE=1` advice in the message.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: backend/agents/investment_team/trading_service/service.py [hunk 2]
```diff
@@ -527,6 +590,278 @@ def run(
...
+
+            # Group orders/cancels by bar_index. Untagged records (legacy
+            # child) fall back to bar 0 — defensive; the new child always
+            # tags chunked records.
+            orders_by_bar: Dict[int, List[Dict]] = {}
+            for o, idx in zip(chunk_resp.orders, chunk_resp.order_bar_indices):
+                orders_by_bar.setdefault(idx if idx is not None else 0, []).append(o)
+            cancels_by_bar: Dict[int, List[Dict]] = {}
+            for c, idx in zip(chunk_resp.cancels, chunk_resp.cancel_bar_indices):
+                cancels_by_bar.setdefault(idx if idx is not None else 0, []).append(c)
```
[the remark this answers is not in the document]
Comment (brandonkindred) [reply backend/agents/investment_team/trading_service/service.py]: Addressed in a756d8e. `_run_chunked` now validates each `bar_index` against `[0, len(chunk_buffer))` before bucketing into `orders_by_bar` / `cancels_by_bar`; out-of-range or `None` indices raise `StrategyRuntimeError(etype="protocol_error")` (which the outer try/except in `_run_chunked` surfaces as a structured `TradingServiceResult.error` with `"out-of-range bar_index"` in the message). No more silent drops via `orders_by_bar.get(i, [])` returning empty.

New regression test: `test_chunked_path_rejects_out_of_range_bar_index` runs a strategy that hand-sets `ctx._current_bar_index = 99` through `run_backtest` with `BAR_CHUNK_SIZE=4` and asserts `result.service_result.error` is populated.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by brandonkindred** [comment]
@codex review and tell me the first 20 issues that you findl

File: backend/agents/investment_team/trading_service/service.py [hunk 3]
```diff
@@ -527,6 +590,296 @@ def run(
...
+            # protocol violation.
+            chunk_len = len(chunk_buffer)
+
+            def _validated(
+                records: List[Dict], indices: List[Optional[int]], kind: str
+            ) -> Dict[int, List[Dict]]:
+                grouped: Dict[int, List[Dict]] = {}
+                for rec, idx in zip(records, indices):
+                    if not isinstance(idx, int) or not (0 <= idx < chunk_len):
+                        raise StrategyRuntimeError(
```
[the remark this answers is not in the document]
Comment (brandonkindred) [reply backend/agents/investment_team/trading_service/service.py]: Good catch — you're right that the in-range validation alone doesn't prevent a strategy from setting `ctx._current_bar_index` after observing later bars and emitting with an in-range earlier index. Addressed in 1ff4063 by **taking bar_index away from strategy code entirely**:

- `contract.StrategyContext.submit_order` / `cancel` no longer attach `bar_index` themselves and `_current_bar_index` is removed from the context. Strategy code has no attribute (public or private) it can write that influences emitted `bar_index`.
- The child harness wraps `_emit` with a `_tagged_emit` closure that captures a private `_chunk_state["i"]` dict (not reachable from `ctx`) and **overrides** `bar_index` on every `order` / `cancel` record before it leaves the subprocess. The dispatch loop sets `_chunk_state["i"] = i` in a try/finally so an exception inside `on_bar` can't leak a stale index.

The parent-side range validation from the previous round stays put as defense-in-depth — with the harness now the sole source of truth, it only fires on genuine protocol bugs.

New regression test `test_chunked_bar_index_cannot_be_forged_by_strategy` runs a fixture strategy that aggressively tries to forge bar_index (sets `ctx._current_bar_index = 99`, `ctx.bar_index = -7` before each emission) and asserts every emitted `bar_index` still matches the harness-managed iteration index in monotonic order.

---
_Generated by [Claude Code](https://claude.ai/code)_
